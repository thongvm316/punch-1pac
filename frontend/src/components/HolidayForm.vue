<template>
  <div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.name.$error || errors.name }"
    >
      <label class="form-label">{{ $t('label.name') }}</label>
      <input
        v-model="$v.params.name.$model"
        class="form-input"
        type="text"
      >
      <p
        v-if="$v.params.name.$error && !errors.name"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.name') }) }}
      </p>
      <p
        v-if="errors.name"
        class="form-input-hint"
      >
        {{ $t('label.name') }} {{ errors.name[0] }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.started_at.$error || errors.started_at }"
    >
      <label class="form-label">{{ $t('label.startAt') }}</label>
      <flat-pickr
        v-model="$v.params.started_at.$model"
        :config="{ mode: 'single', locale: flatpickrLocaleMapper[pickrLocale] }"
        class="form-input daterange-picker"
      />
      <p
        v-if="errors.started_at"
        class="form-input-hint"
      >
        {{ $t('label.startAt') }} {{ errors.started_at[0] }}
      </p>
      <p
        v-if="$v.params.started_at.$error && !errors.started_at"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.startAt') }) }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.ended_at.$error || errors.ended_at }"
    >
      <label class="form-label">{{ $t('label.endAt') }}</label>
      <flat-pickr
        v-model="$v.params.ended_at.$model"
        :config="{ mode: 'single', locale: flatpickrLocaleMapper[pickrLocale] }"
        class="form-input daterange-picker"
      />
      <p
        v-if="errors.ended_at"
        class="form-input-hint"
      >
        {{ $t('label.endAt') }} {{ errors.ended_at[0] }}
      </p>
      <p
        v-if="$v.params.ended_at.$error && !errors.ended_at"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.endAt') }) }}
      </p>
    </div>
    <div class="form-group">
      <button
        v-if="!targetHoliday"
        ref="localAddHolidayBtn"
        class="btn btn-success btn-submit"
        :disabled="isDisabled"
        @click="localAddHoliday"
      >
        {{ $t('button.common.submit') }}
      </button>
      <button
        v-if="targetHoliday"
        ref="localEditHolidayBtn"
        class="btn btn-success btn-submit"
        :disabled="isDisabled"
        @click="localEditHoliday"
      >
        {{ $t('button.common.save') }}
      </button>
    </div>
  </div>
</template>

<script>
import flatpickrLocale from '../mixins/flatpickr-locale'
import handleSuccess from '../mixins/handle-success'
import holidayFormValidate from '../validations/holiday-form-validate'
import { CLEAR_HOLIDAY_ERRORS } from '../store/mutation-types'
import { mapState, mapActions, mapMutations } from 'vuex'
import { isEmpty } from 'underscore'
const flatPickr = () => import('vue-flatpickr-component')

export default {
  name: 'HolidayForm',

  components: {
    flatPickr
  },

  mixins: [flatpickrLocale, handleSuccess, holidayFormValidate],

  props: {
    targetHoliday: {
      type: Object,
      default: null
    }
  },

  data() {
    return {
      params: {
        name: '',
        started_at: '',
        ended_at: ''
      }
    }
  },

  computed: {
    ...mapState('companyHolidays', ['errors'])
  },

  created() {
    if (!isEmpty(this.errors)) this[CLEAR_HOLIDAY_ERRORS]()
    if (this.targetHoliday) this.params = { ...this.targetHoliday }
  },

  methods: {
    ...mapActions('companyHolidays', ['createHoliday', 'updateHoliday']),

    ...mapMutations('companyHolidays', [CLEAR_HOLIDAY_ERRORS]),

    localAddHoliday() {
      this.createHoliday(this.params).then(response => {
        Object.keys(this.params).forEach(key => {
          this.params[key] = ''
        })
        this.handleSuccess({
          emitType: 'afterModify',
          message: this.$t('messages.holiday.createSuccess')
        })
      })
    },

    localEditHoliday() {
      this.updateHoliday({ holidayID: this.targetHoliday.id, updateParams: this.params }).then(response => {
        this.handleSuccess({
          emitType: 'afterModify',
          message: this.$t('messages.holiday.updateSuccess')
        })
      })
    }
  }
}
</script>
