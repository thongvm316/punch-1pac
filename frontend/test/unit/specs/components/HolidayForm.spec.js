import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import setComputed from '../../supports/set-computed'
import holidaysData from '../../supports/fixtures/holidays.api'
import HolidayForm from '@/components/HolidayForm'

const localAddHoliday = jest.spyOn(HolidayForm.methods, 'localAddHoliday')
const localEditHoliday = jest.spyOn(HolidayForm.methods, 'localEditHoliday')
const CLEAR_HOLIDAY_ERRORS = jest.spyOn(HolidayForm.methods, 'CLEAR_HOLIDAY_ERRORS')
const createHoliday = jest.spyOn(HolidayForm.methods, 'createHoliday').mockResolvedValue(null)
const updateHoliday = jest.spyOn(HolidayForm.methods, 'updateHoliday').mockResolvedValue(null)
const handleSuccess = jest.spyOn(HolidayForm.mixins[1].methods, 'handleSuccess')

const targetHoliday = [...holidaysData.holidays][0]

const localWrapperOps = {
  ...wrapperOps,
  stubs: {
    flatPickr: true
  },
  computed: {
    isDisabled: () => false,
    pickLocale: () => 'en'
  }
}

describe('HolidayForm.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(HolidayForm, localWrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when have no props', () => {
    describe('when HolidayForm was mounted', () => {
      it('should render correctly', () => {
        expect(wrapper.exists()).toBeTruthy()
        expect(wrapper.isVueInstance()).toBeTruthy()
        expect(wrapper).toMatchSnapshot()
      })
    })

    describe('when localAddHoliday', () => {
      it('should localAddHoliday hasBeenCalled', async () => {
        wrapper.find({ ref: 'localAddHolidayBtn' }).trigger('click')
        await wrapper.vm.$nextTick()

        expect(localAddHoliday).toHaveBeenCalled()
        expect(createHoliday).toHaveBeenCalledWith(wrapper.vm.params)
        expect(handleSuccess).toHaveBeenCalledWith({ emitType: 'afterModify', message: 'Holiday is created' })
      })
    })
  })

  describe('when have props', () => {
    beforeEach(async () => {
      wrapper.setProps({
        targetHoliday
      })
      await wrapper.vm.$nextTick()
    })

    describe('when HolidayForm was mounted', () => {
      it('should render correctly', () => {
        expect(wrapper.exists()).toBeTruthy()
        expect(wrapper.isVueInstance()).toBeTruthy()
        expect(wrapper).toMatchSnapshot()
      })
    })

    describe('when localEditHoliday', () => {
      it('should localEditHoliday haveBeenCalled', async () => {
        wrapper.find({ ref: 'localEditHolidayBtn' }).trigger('click')
        await wrapper.vm.$nextTick()

        expect(localEditHoliday).toHaveBeenCalled()
        expect(updateHoliday).toHaveBeenCalledWith({
          holidayID: targetHoliday.id,
          updateParams: wrapper.vm.params
        })
        expect(handleSuccess).toHaveBeenCalledWith({ emitType: 'afterModify', message: 'Holiday is updated' })
      })
    })
  })

  describe('when errors', () => {
    it('should have text error', () => {
      const errors = {
        name: ['has been taken'],
        started_at: ['wrong'],
        ended_at: ['wrong']
      }
      setComputed(wrapper, { errors })

      expect(wrapper).toMatchSnapshot()
    })
  })
})
