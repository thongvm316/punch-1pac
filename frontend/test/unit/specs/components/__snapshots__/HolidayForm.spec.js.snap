// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HolidayForm.vue when errors should have text error 1`] = `
<div>
  <div class="form-group has-error"><label class="form-label">Name</label> <input type="text" class="form-input">
    <!---->
    <p class="form-input-hint">
      Name has been taken
    </p>
  </div>
  <div class="form-group has-error"><label class="form-label">Start at</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <p class="form-input-hint">
      Start at wrong
    </p>
    <!---->
  </div>
  <div class="form-group has-error"><label class="form-label">End at</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <p class="form-input-hint">
      End at wrong
    </p>
    <!---->
  </div>
  <div class="form-group"><button class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`HolidayForm.vue when have no props when HolidayForm was mounted should render correctly 1`] = `
<div>
  <div class="form-group"><label class="form-label">Name</label> <input type="text" class="form-input">
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Start at</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">End at</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`HolidayForm.vue when have props when HolidayForm was mounted should render correctly 1`] = `
<div>
  <div class="form-group"><label class="form-label">Name</label> <input type="text" class="form-input">
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Start at</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">End at</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group">
    <!----> <button class="btn btn-success btn-submit">
      Save
    </button></div>
</div>
`;
